### PR TITLE
feature/domain_filtering_for_email_sources

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Allow to filter email domains that can be used for the email source

Impacts
=======
EmailSource and the portal email source workflow

Checklist
=======
- [x] Document the feature
- [X] Add unit tests
- [x] Add acceptance tests (TestLink)

NEWS file entries
=================

Enhancements
------------
* Email Source can have banned and allowed email domains

### CODE REVIEW

File: html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm [hunk 1]
```diff
@@ -51,6 +51,18 @@ has_field 'activation_domain' =>
     },
   );
 
+has_field 'allowed_domains' =>
+  (
+   'type' => 'Text',
+    label => 'Comma seperated list of Allowed Domains',
```
Comment (extrafu) [inline html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm]: Comma-separated

File: html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm [hunk 2]
```diff
@@ -51,6 +51,18 @@ has_field 'activation_domain' =>
...
+  );
+
+has_field 'banned_domains' =>
+  (
+   'type' => 'Text',
+    label => 'Comma seperated list of Banned Domains',
```
Comment (extrafu) [inline html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm]: Comma-separated

**Comment by extrafu** [comment]
@nqb Please test it.

File: html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm (hunk 1 above)
Comment (nqb) [reply html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm]: seperated -> separated

File: html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm (hunk 2 above)
Comment (nqb) [reply html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm]: seperated -> separated

**Comment by nqb** [comment]
@extrafu: I didn't see any calls to `isEmailAllowed` method except for unit tests, are you sure the PR is ready for tests ?

I try to add one domain to the allowed domains field on default email source, I got following error: https://gist.github.com/nqb/c6479592b3d48107cbae3887e7876a9a

Few remarks:
- Feature has been requested to be use on **sponsor** sources
- We should increase size box of the new fields:
screenshot from 2018-11-22 13-32-07

**Comment by nqb** [comment]
I updated description of the PR to add the new checklist. 

@jrouzierinverse: you should update the way you create PR based on https://github.com/inverse-inc/packetfence/blob/8e1d98af80a101a45a49528c6e465a85d55477da/.github/PULL_REQUEST_TEMPLATE.md#checklist

**Review by nqb — CHANGES_REQUESTED** [review]
Please add comments for `isEmailAllowed`, `domains_regexes` and `make_regex` methods.

**Comment by jrouzierinverse** [comment]
Sorry did not push the commit

**Comment by jrouzierinverse** [comment]
Done

**Comment by jrouzierinverse** [comment]
Ready for testing.
Will update documentation

**Comment by nqb** [comment]
If I put domains in fields, it works but when I get back to my source, domain fields looks like this: `ARRAY(0x558e032f40c0)` but seems fine in `authentication.conf`.

File: lib/pf/Authentication/Source/EmailSource.pm [hunk 3]
```diff
@@ -104,20 +107,85 @@ sub mandatoryFields {
 sub authenticate {
     my ( $self, $username, $password ) = @_;
     my $logger = pf::log::get_logger;
+    if (!$self->isEmailAllowed($username)) {
+        $logger->warn("Failed to authenticate using EmailSource ($self->{id}) with PID '$username'");
```
Comment (nqb) [inline lib/pf/Authentication/Source/EmailSource.pm]: IMO, we should give a more precise reason in log, like you did before.
Is it possible to include content of "EMAIL_UNAUTHORIZED" ?

Comment (jrouzierinverse) [reply lib/pf/Authentication/Source/EmailSource.pm]: Of course we can use the include the contents of 'EMAIL_UNAUTHORIZED'
What more precise information to you need?
I included the source id and email address in the log what else do you think we need?

Comment (nqb) [reply lib/pf/Authentication/Source/EmailSource.pm]: Just to say that the authentication failed due to "banned" domain

**Comment by nqb** [comment]
No more `ARRAY(0x...)` in fields.

File: html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm [hunk 4]
```diff
@@ -51,6 +51,26 @@ has_field 'activation_domain' =>
...
+  );
+
+has_field 'banned_domains' =>
+  (
+    type  => 'TextArea',
+    label => 'Comma-separated list of Banned Domains',
+    tags => {
+        after_element => \&help,
+        help => 'A comma-separated list of domains that are banned for email registration. Banned domains are checked before allowed domains.'
+    },
```
Comment (nqb) [inline html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm]: if banned domains are checked before allowed domains, I propose to reverse order of fields in the form.

File: lib/pf/Authentication/Source/EmailSource.pm [hunk 5]
```diff
@@ -103,21 +134,86 @@ sub mandatoryFields {
...
+    local $_;
+    if ($d =~ /\*/) {
+        $d =~ s/\./\\./g;
+        $d =~ s/\*/\.\*/g;
+        $d = "\@$d\$";
+        return qr/$d/;
+    }
+
+    return qr/@\Q$d\E$/;
+}
```
Comment (nqb) [inline lib/pf/Authentication/Source/EmailSource.pm]: Could you describe what check the if and what return the function ?

File: lib/pf/constants/authentication/messages.pm [hunk 6]
```diff
@@ -17,14 +17,15 @@ use warnings;
...
-  $COMMUNICATION_ERROR_MSG $AUTH_FAIL_MSG $AUTH_SUCCESS_MSG $INVALID_EMAIL_MSG $LOCALDOMAIN_EMAIL_UNAUTHORIZED
+  $COMMUNICATION_ERROR_MSG $AUTH_FAIL_MSG $AUTH_SUCCESS_MSG $INVALID_EMAIL_MSG $LOCALDOMAIN_EMAIL_UNAUTHORIZED $EMAIL_UNAUTHORIZED
 );
 
 Readonly our $COMMUNICATION_ERROR_MSG => 'Unable to validate credentials at the moment';
 Readonly our $AUTH_FAIL_MSG => 'Invalid login or password';
 Readonly our $AUTH_SUCCESS_MSG => 'Authentication successful.';
 Readonly our $INVALID_EMAIL_MSG => 'Invalid e-mail address';
 Readonly our $LOCALDOMAIN_EMAIL_UNAUTHORIZED => "You can't register as a guest with this corporate email address. Please register as a regular user using your email address instead.";
+Readonly our $EMAIL_UNAUTHORIZED => "Cannot register with this email address.";
```
Comment (nqb) [inline lib/pf/constants/authentication/messages.pm]:
```suggestion
Readonly our $EMAIL_UNAUTHORIZED => "Cannot register with this email address: unallowed domain.";
```

File: html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm (hunk 4 above)
Comment (nqb) [reply html/pfappserver/lib/pfappserver/Form/Config/Source/Email.pm]: If we support regex in domains fields, we should update "help" statements.

**Review by nqb — CHANGES_REQUESTED** [review]

**Comment by jrouzierinverse** [comment]
Done re-review

**Comment by nqb** [comment]
Thanks a lot for your changes.

My new remarks:
- register using local domain in uppercase work, not for allowed/banned domains. We need to be consistent: it should not work in any cases.
- as said before, feature has been requested to be use on sponsor sources. We need to port it.

**Comment by nqb** [comment]
I created acceptance tests.

**Comment by jrouzierinverse** [comment]
Did the issue come up when the allowed/banned domains are uppercase in the config or during the authentication?

**Comment by nqb** [comment]
I'm able to reproduce the issue with following config:
- allow local domain: checked
- banned domains: empty
- allowed domains: one uppercase domain (FREE.fr)

My local domain is: centos-devel.example.lan, I'm able to register with nqb@CENTOS-devel.example.lan. No changes if I put my allowed domain in lowercase.

**Comment by nqb** [comment]
> register using local domain in uppercase work, not for allowed/banned domains. We need to be consistent: it should not work in any cases.

Still presents.

I got a new issue. If I edit the default email and sponsor sources and update only the `Comma-separated list of Allowed Domains` field with one domain, I got this [error](https://gist.github.com/nqb/27a40353c8fe54b97aeb1ed837364dc3) in logs.

If I filled the `Comma-separated list of Banned Domains` field, error disappears.

**Comment by jrouzierinverse** [comment]
Done

**Comment by nqb** [comment]
@jrouzierinverse : is this code can be apply as is on a 8.1 installation ?

**Comment by nqb** [comment]
Issue in webadmin fix, I don't see the fix for register using local domain in uppercase

**Comment by jrouzierinverse** [comment]
Here is the fix dfcc31c08fc6f92486bf5aec1f8be2c4dd330d40 and test 2e1b457d491e1d8d6ac250e4adccbff2d050af19.

**Comment by nqb** [comment]
Now, I'm able to register with domain in uppercase or lowercase.

**Review by nqb — APPROVED** [review]

**Comment by extrafu** [comment]
@nqb If all is fine, please merge and update the NEWS file

**Pull request merged**